Cleanup MAC

levitte opened the pull request:
Two things are happening in this changeset:

## Remove MAC cruft

A few declarations and static functions / arrays that are no longer
used were left behind when MACs moved to be implemented by providers.

## Get rid of the diversity of names for MAC parameters

The EVP_PKEY MAC implementations had a diversity of controls that were
really the same thing.  We did reproduce that for the provider based
MACs, but are changing our minds on this.  Instead of that, we now use
one parameter name for passing the name of the underlying algorithm to
a MAC implementation, "algorithm", and one parameter name for passing
the output size of the MAC, "size".

Then we leave it to the EVP_PKEY->EVP_MAC bridge to translate "md",
"digest" and "cipher" to "algorithm", and "digestsize" to "size".

levitte committed:
Remove MAC cruft

A few declarations and static functions / arrays that are no longer
used were left behind when MACs moved to be implemented by providers.

levitte committed:
Get rid of the diversity of names for MAC parameters

The EVP_PKEY MAC implementations had a diversity of controls that were
really the same thing.  We did reproduce that for the provider based
MACs, but are changing our minds on this.  Instead of that, we now use
one parameter name for passing the name of the underlying algorithm to
a MAC implementation, "algorithm", and one parameter name for passing
the output size of the MAC, "size".

Then we leave it to the EVP_PKEY->EVP_MAC bridge to translate "md",
"digest" and "cipher" to "algorithm", and "digestsize" to "size".

slontis commented on crypto/evp/pkey_mac.c:
```diff
@@ -441,6 +441,24 @@ static int pkey_mac_ctrl_str(EVP_PKEY_CTX *ctx,
+     * "digestsize" was a setting control in siphash, but naming wise,
+     * it's really the same as "size".
+     */
+    if (strcmp(type, "md") == 0
```
Are you sure about this.. 
In kdf at least there can be two algorithms so this wont work for that case.. I dont know if I like this.

slontis replied:
I guess what I am saying is that it is going to be inconsistent with KDF.

levitte replied:
The only I can find is sskdf, which can have a MAC and an MD.  So it may make sense to have "digest", "cipher" and "mac" for KDFs...  and yeah, in the interest of not having too much diversity, it might be suitable to have "digest" and "cipher" for MACs as well, then.

Do you agree, though, that "md" and "digest" can be merged into one, i.e. one gets translated into the other (I would rather standardise on "digest")?

slontis replied:
Yes I agree..

levitte replied:
Cool, that's a good enough compromise for me :-)

slontis commented on include/openssl/core_names.h:
```diff
@@ -75,18 +75,14 @@ extern "C" {
 #define OSSL_MAC_PARAM_ENGINE       "engine"    /* utf8 string */
 #define OSSL_MAC_PARAM_PROPERTIES   "properties" /* utf8 string */
-/* Note that "size", "digestsize" and "outlen" are equivalent */
 #define OSSL_MAC_PARAM_SIZE         "size"      /* size_t */
```
None of these is an input size?

levitte replied:
Not as far as I know...  The only that actually existed in the old EVP_PKEY method form was "digestsize".  "size" was named such to correspond to the `EVP_MAC_size()` function, and "outlen" was my invention because I thought it was more expressive than "size", but I've realised that it would be an uphill battle to have that adopted, so I have it up.

levitte committed:
squash! Get rid of the diversity of names for MAC parameters

After some discussion, @slontis and I agreed to keep "digest" and
"cipher", because that would be needed for KDFs anyway so might as
well keep things consistent.  "md" still goes away in favor of
"digest", and "algorithm" goes away period.

levitte committed:
Update EVP_MAC.pod

levitte commented:
Reworked a bit following your remarks, @slontis.  I believe I addressed them all

paulidale commented:
I've changed the KDFs/PRFs draft to match this.

slontis commented on crypto/evp/m_sha3.c:
```diff
@@ -36,12 +36,6 @@ static int shake_init(EVP_MD_CTX *ctx)
     return sha3_init(EVP_MD_CTX_md_data(ctx), '\x1f', ctx->digest->md_size * 8);
 }
 
-static int kmac_init(EVP_MD_CTX *ctx)
```
So this has been removed because this is the legacy code area, and kmac is new to 3.0.0 so it only belongs in the provider?

levitte replied:
It was only used by `evp_keccak_kmac128()` and `evp_keccak_kmac256()`, which aren't used any more because the provider KMAC implementations use the provider KECCAK_KMAK implementations.  In other words, dead code, and also entirely internal, so regardless of if it's new or not, it is harmless to remove

slontis commented on doc/man7/provider-mac.pod:
```diff
@@ -178,21 +178,12 @@ Gets flags associated with the MAC.
-the expected operation for the underlying algorithm.
-These are regarded as antiquated, but are kept for easier transition from
-legacy MAC implementations.
+Sets the name of the underlying cipher or digest to be used.
```
This probably needs to be seperated now so that param_cipher has its own description.

levitte replied:
Why?  isn't "underlying cipher or digest" clear enough?

slontis replied:
I guess its ok..

slontis reviewed (approved):
Looks good.

levitte commented:
Cool, thanks!

Merged.

9f57e2184de330e83d0636586e88a8524c513b8e Remove MAC cruft
703170d4b9da1b3adc0e4d87719a5156080b2dca Get rid of the diversity of names for MAC parameters
f3b8d77f99c84e39c48599ad54366e806f3b9478 Update EVP_MAC.pod
